src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties': {
+        Object.assign(stateCopy, action.extraData);
+        break;
+      }
+
+      case 'removeProperties': {
+        action.keysToRemove.forEach(key => {
+          delete stateCopy[key];
+        });
+        break;
+      }
+
+      case 'clear':
+        stateCopy = {};
+        break;
+
+      default:
+        return `Something went wrong, there's no '${action.type}' action type`;
```

Comment:
But better would be to [throw an error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/throw#description)